edgelet/edgelet-docker/src/runtime.rs

```diff
@@ -829,6 +777,66 @@ where
     )
 }
 
+fn authenticate<MR>(
+    runtime: &MR,
+    req: &Request<Body>,
+) -> Box<dyn Future<Item = AuthId, Error = Error> + Send>
+where
+    MR: ModuleRuntime<Error = Error>,
+    <MR as ModuleRuntime>::ListFuture: 'static,
+    MR::Module: DockerModuleTop<Error = Error>,
+{
+    let pid = req
+        .extensions()
+        .get::<Pid>()
+        .cloned()
+        .unwrap_or_else(|| Pid::None);
+
+    Box::new(match pid {
+        Pid::None => Either::A(future::ok(AuthId::None)),
+        Pid::Any => Either::A(future::ok(AuthId::Any)),
+        Pid::Value(pid) => {
+            Either::B(
+                runtime
+                    .list()
+                    .into_stream()
+                    .map(|list| {
+                        stream::futures_unordered(list.into_iter().map(|module| module.top()))
+                    })
+                    .flatten()
+                    .then(Ok::<_, Error>) // wrap error into a result to ignore Module::NotFound error
+                    .filter_map(move |top| match top {
+                        Ok(top) => {
+                            if top.process_ids().contains(&pid) {
+                                Some(Ok(top.name().to_string()))
+                            } else {
+                                None
+                            }
+                        }
+                        Err(err) => match err.kind() {
+                            ErrorKind::NotFound(_)
+                            | ErrorKind::RuntimeOperation(RuntimeOperation::TopModule(_)) => None,
+                            _ => Some(Err(err)),
+                        },
+                    })
+                    .then(Result::unwrap) // unwrap unknown error to emit error
```

Comment:
If the Result::unwrap fails, will this panic or is it handled by the code below?

Reply:
It should not be the case here. On line 807 I wrapped everything in Result to filter out known errors, that for authentication purposes will be ignored.